src/main/kotlin/org/radarbase/push/integration/garmin/service/BackfillService.kt

```diff
@@ -11,9 +28,86 @@ import javax.ws.rs.core.Context
  * service to create POST requests for historic data to our server.
  */
 class BackfillService(
-    @Context config: Config,
-    @Named(GARMIN_QUALIFIER) userRepository: GarminUserRepository
-) {
-    // TODO: discuss about this since we need some way to store the offset for historic requests
-    //  made for each user
+    @Context private val config: Config,
+    @Named(GARMIN_QUALIFIER) private val userRepository: GarminUserRepository
+) : ApplicationEventListener {
+
+    private val redisHolder =
+        RedisHolder(JedisPool(config.pushIntegration.garmin.backfill.redis.uri))
+    private val executorService = Executors.newSingleThreadScheduledExecutor()
+    private val requestExecutorService = Executors.newFixedThreadPool(
+        config.pushIntegration.garmin.backfill.maxThreads
+    )
+    private val requestGenerator = GarminRequestGenerator(config, userRepository)
+    private val remoteLockManager = RedisRemoteLockManager(
+        redisHolder,
+        config.pushIntegration.garmin.backfill.redis.lockPrefix
+    )
+    private val httpClient = OkHttpClient()
+    private val requestsPerUserPerIteration: Int
+        get() = 40
+
+    private val futures: MutableList<Future<*>> = mutableListOf()
+
+    override fun onEvent(event: ApplicationEvent?) {
+        when (event?.type) {
+            INITIALIZATION_FINISHED -> start()
+            DESTROY_FINISHED -> stop()
+            else -> logger.info("Application event received: ${event?.type}")
+        }
+    }
+
+    override fun onRequest(requestEvent: RequestEvent?): RequestEventListener? = null
+
+    private fun start() {
+        logger.info("Application Initialisation completed. Starting Backfill service...")
+
+        executorService.scheduleAtFixedRate(::iterateUsers, 1, 5, TimeUnit.MINUTES)
+    }
+
+    private fun stop() {
+        logger.info("Application Destroy completed. Stopping Backfill service...")
+        try {
+            requestExecutorService.awaitTermination(30, TimeUnit.SECONDS)
+            executorService.awaitTermination(30, TimeUnit.SECONDS)
+        } catch (e: InterruptedException) {
+            logger.error("Failed to complete execution: interrupted")
+        }
+    }
+
+    private fun iterateUsers() {
+        while (futures.any { !it.isDone }) {
+            logger.info("The previous task is already running. Waiting ${WAIT_TIME_MS / 1000}s...")
+            Thread.sleep(WAIT_TIME_MS)
+        }
+        futures.clear()
+        logger.info("Making Garmin Backfill requests...")
+        try {
+            userRepository.stream().forEach { user ->
+                remoteLockManager.tryRunLocked(user.versionedId) {
+                    requestGenerator.requests(user, requestsPerUserPerIteration).forEach { req ->
+                        futures.add(requestExecutorService.submit { makeRequest(req) })
+                    }
+                }
```

Comment:
I'm not sure what the purpose of the lock is. In any case, the lock does not apply to anything scheduled inside another thread, so the actual request will not be locked. Is that intentional? Otherwise, there should just be a lock around updating the offsets.

Reply:
The lock is just for specifying that currently requests are being made for the current user, so another instance does not make the exact same requests. Will that work this way?

Reply:
No, it will only lock while the requests are generated, but it will not lock for simultaneous requests being made for the user. Since the scheduler thread is already running iterateUsers, perhaps the additional futures.add(requestExecutorService.submit) is not needed? If you want to keep this more granular as it currently is, you could rewrite it to

```kotlin
futures.add(requestExecutorService.submit {
    remoteLockManager.tryRunLocked(user.versionedId) {
        requestGenerator.requests(user, requestsPerUserPerIteration).forEach { req ->
            makeRequest(req)
        }
    }
})
```

Since you don't expect the lock to overlap even though it could happen, the overhead of failed locks is quite low. The lock also encompasses all of the code it should lock.